File: apps/android/app/src/main/java/com/pcmobilelink/nearshare/pairing/PairingOfferDiscoveryClient.kt
package com.pcmobilelink.nearshare.pairing

import com.pcmobilelink.nearshare.discovery.NearShareDiscoveryProtocol
import com.pcmobilelink.nearshare.discovery.UdpBroadcastAddresses
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.SocketTimeoutException
import org.json.JSONObject

class PairingOfferDiscoveryClient(
    private val timeoutMillis: Int = 15_000,
) {
    fun resolve(shortCode: String): PairingPayload {
        val normalized = PairingShortCode.normalize(shortCode)
        require(PairingShortCode.isValid(normalized)) {
            "Enter the 9-character pairing code shown on the other device."
        }

        DatagramSocket().use { socket ->
            socket.broadcast = true
            socket.soTimeout = BroadcastIntervalMillis

            val request = JSONObject()
                .put("type", AndroidPairingOfferDiscoveryResponder.REQUEST_TYPE)
                .put("shortCode", normalized)
                .toString()
                .toByteArray(Charsets.UTF_8)

            val buffer = ByteArray(MAX_DATAGRAM_BYTES)
            val deadline = System.currentTimeMillis() + timeoutMillis
            while (System.currentTimeMillis() < deadline) {
                broadcastRequest(socket, request)
                while (true) {
                    try {
                        val packet = DatagramPacket(buffer, buffer.size)
                        socket.receive(packet)
                        parseResponse(
                            responseBytes = packet.data.copyOf(packet.length),
                            expectedShortCode = normalized,
                            responderHost = packet.address.hostAddress,
                        )?.let { return it }
                    } catch (_: SocketTimeoutException) {
                        break
                    }
                }
            }

            throw IllegalStateException("Could not find that pairing code. Join the same Wi-Fi, or create and join a private connection first.")
        }
    }

    private fun broadcastRequest(socket: DatagramSocket, request: ByteArray) {
        UdpBroadcastAddresses.resolve().forEach { address ->
            runCatching {
                socket.send(
                    DatagramPacket(
                        request,
                        request.size,
                        address,
                        NearShareDiscoveryProtocol.DISCOVERY_PORT,
                    ),
                )
            }
        }
    }

    private fun parseResponse(
        responseBytes: ByteArray,
        expectedShortCode: String,
        responderHost: String?,
    ): PairingPayload? {
        return runCatching {
            val json = JSONObject(responseBytes.toString(Charsets.UTF_8))
            if (json.optString("type") != AndroidPairingOfferDiscoveryResponder.RESPONSE_TYPE) {
                return null
            }
            if (PairingShortCode.normalize(json.optString("shortCode")) != expectedShortCode) {
                return null
            }

            val pairingUri = json.optString("pairingUri")
            if (pairingUri.isBlank()) {
                return null
            }

            val payload = PairingPayloadCodec.decode(pairingUri)
            if (PairingShortCode.normalize(payload.shortCode.orEmpty()) != expectedShortCode) {
                return null
            }
            payload.preferResponderEndpoint(responderHost)
        }.getOrNull()
    }

    private fun PairingPayload.preferResponderEndpoint(responderHost: String?): PairingPayload {
        val primaryEndpoint = endpoints.firstOrNull()
        if (primaryEndpoint == null || responderHost.isNullOrBlank()) {
            return this
        }

        val responderEndpoint = PairingEndpointCandidate(
            host = responderHost,
            port = primaryEndpoint.port,
        )
        val preferredEndpoints = buildList {
            add(responderEndpoint)
            endpoints.forEach { endpoint ->
                if (!endpoint.host.equals(responderEndpoint.host, ignoreCase = true) ||
                    endpoint.port != responderEndpoint.port
                ) {
                    add(endpoint)
                }
            }
        }
        return copy(endpoints = preferredEndpoints)
    }

    private companion object {
        private const val BroadcastIntervalMillis = 750
        private const val MAX_DATAGRAM_BYTES = 8192
    }
}

File: apps/windows/src/PCMobileLink.Windows.Infrastructure/Pairing/PairingOfferDiscoveryClient.cs
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Text.Json;
using PCMobileLink.Windows.Core;
using PCMobileLink.Windows.Infrastructure.Discovery;

namespace PCMobileLink.Windows.Infrastructure.Pairing;

public sealed class PairingOfferDiscoveryClient
{
    private const string RequestType = "nearshare.pairing.discovery.request.v1";
    private const string ResponseType = "nearshare.pairing.discovery.response.v1";
    private static readonly TimeSpan BroadcastInterval = TimeSpan.FromMilliseconds(750);
    private static readonly JsonSerializerOptions SerializerOptions = new(JsonSerializerDefaults.Web);
    private readonly TimeSpan _timeout;

    public PairingOfferDiscoveryClient(TimeSpan? timeout = null)
    {
        _timeout = timeout ?? TimeSpan.FromSeconds(15);
    }

    public async Task<PairingPayload> ResolveAsync(string shortCode, CancellationToken cancellationToken = default)
    {
        string normalized = PairingShortCode.Normalize(shortCode);
        if (!PairingShortCode.IsValid(normalized))
        {
            throw new FormatException("Enter the 9-character pairing code shown on the other device.");
        }

        using UdpClient udpClient = new(AddressFamily.InterNetwork)
        {
            EnableBroadcast = true
        };
        udpClient.Client.Bind(new IPEndPoint(IPAddress.Any, 0));

        byte[] requestBytes = JsonSerializer.SerializeToUtf8Bytes(
            new PairingOfferDiscoveryRequest
            {
                Type = RequestType,
                ShortCode = normalized
            },
            SerializerOptions);

        DateTimeOffset deadline = DateTimeOffset.UtcNow.Add(_timeout);
        while (DateTimeOffset.UtcNow < deadline)
        {
            cancellationToken.ThrowIfCancellationRequested();
            await BroadcastRequestAsync(udpClient, requestBytes).ConfigureAwait(false);

            TimeSpan remaining = deadline - DateTimeOffset.UtcNow;
            if (remaining <= TimeSpan.Zero)
            {
                break;
            }

            PairingPayload? payload = await TryReceivePayloadAsync(
                    udpClient,
                    normalized,
                    remaining < BroadcastInterval ? remaining : BroadcastInterval,
                    cancellationToken)
                .ConfigureAwait(false);
            if (payload is not null)
            {
                return payload;
            }
        }

        throw new TimeoutException("Could not find that pairing code. Join the same Wi-Fi, or create and join a private connection first.");
    }

    private static async Task BroadcastRequestAsync(UdpClient udpClient, byte[] requestBytes)
    {
        foreach (IPEndPoint endpoint in UdpBroadcastEndpoints.Resolve(LocalDiscoveryResponderOptions.DefaultDiscoveryPort))
        {
            try
            {
                await udpClient.SendAsync(requestBytes, requestBytes.Length, endpoint).ConfigureAwait(false);
            }
            catch (SocketException)
            {
                // Some adapters reject directed broadcasts; keep trying the remaining local interfaces.
            }
        }
    }

    private static async Task<PairingPayload?> TryReceivePayloadAsync(
        UdpClient udpClient,
        string normalized,
        TimeSpan receiveWindow,
        CancellationToken cancellationToken)
    {
        using CancellationTokenSource timeoutSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        timeoutSource.CancelAfter(receiveWindow);

        while (!timeoutSource.IsCancellationRequested)
        {
            UdpReceiveResult result;
            try
            {
                result = await udpClient.ReceiveAsync(timeoutSource.Token).ConfigureAwait(false);
            }
            catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
            {
                break;
            }

            PairingOfferDiscoveryResponse? response = TryParseResponse(result.Buffer, normalized);
            if (response is null)
            {
                continue;
            }

            PairingPayload payload = PairingPayloadCodec.Decode(response.PairingUri);
            if (!string.Equals(PairingShortCode.Normalize(payload.ShortCode ?? string.Empty), normalized, StringComparison.Ordinal))
            {
                continue;
            }

            return PreferResponderEndpoint(payload, result.RemoteEndPoint.Address);
        }

        return null;
    }

    private static PairingOfferDiscoveryResponse? TryParseResponse(byte[] responseBytes, string expectedShortCode)
    {
        try
        {
            PairingOfferDiscoveryResponse? response = JsonSerializer.Deserialize<PairingOfferDiscoveryResponse>(
                Encoding.UTF8.GetString(responseBytes),
                SerializerOptions);
            if (response is null
                || !string.Equals(response.Type, ResponseType, StringComparison.Ordinal)
                || !string.Equals(PairingShortCode.Normalize(response.ShortCode), expectedShortCode, StringComparison.Ordinal)
                || string.IsNullOrWhiteSpace(response.PairingUri))
            {
                return null;
            }

            return response;
        }
        catch (JsonException)
        {
            return null;
        }
        catch (ArgumentException)
        {
            return null;
        }
    }

    private static PairingPayload PreferResponderEndpoint(PairingPayload payload, IPAddress responderAddress)
    {
        PairingEndpointCandidate? primaryEndpoint = payload.Endpoints.FirstOrDefault();
        if (primaryEndpoint is null
            || IPAddress.Any.Equals(responderAddress)
            || IPAddress.IPv6Any.Equals(responderAddress)
            || IPAddress.None.Equals(responderAddress))
        {
            return payload;
        }

        string responderHost = responderAddress.ToString();
        PairingEndpointCandidate responderEndpoint = new(responderHost, primaryEndpoint.Port);
        List<PairingEndpointCandidate> endpoints = [responderEndpoint];
        endpoints.AddRange(payload.Endpoints.Where(endpoint =>
            endpoint.Port != responderEndpoint.Port
            || !string.Equals(endpoint.Host, responderEndpoint.Host, StringComparison.OrdinalIgnoreCase)));

        return payload with { Endpoints = endpoints };
    }

    private sealed record PairingOfferDiscoveryRequest
    {
        public required string Type { get; init; }

        public required string ShortCode { get; init; }
    }

    private sealed record PairingOfferDiscoveryResponse
    {
        public required string Type { get; init; }

        public required string ShortCode { get; init; }

        public required string DeviceName { get; init; }

        public required string PairingUri { get; init; }

        public required long ServerTimeUnixSeconds { get; init; }
    }
}
